商品購入機能

DDT1129 opened the pull request:
# What

①orderモデル等（商品購入記録）の作成と設定
②Formオブジェクトの利用のための設定、同モデルの単体テストの作成
③API導入、クレジットカード決済での購入機能を実装
④商品販売状況に応じた条件分岐の設定

# Why

①ユーザー、商品情報と紐付けて購入記録データを保存するため
②一つのフォームからの送信で複数モデルを操作でき、またバリデーションを明確にし、可読性の向上、モデルの肥大化防止等のため
③オンラインのクレジットカード決済サービスであるPAYJPを利用することでカード情報をトークン化し、セキュリティ対策をするため
④商品の選択、購入が適正に行われるように各ページのビューを表示するため

# 確認の動画写真（gyazo）

１
必要な情報を適切に入力して「購入」ボタンを押すと、商品の購入ができる動画
https://gyazo.com/bee328b3cefcd306ee759adbe03b0182
 
２
入力に問題がある状態で「購入」ボタンが押された場合、情報は受け入れられず、購入ページでエラーメッセージが表示される動画
https://gyazo.com/00b483163a66bd81a593784831734344
 
３
ログイン状態の場合でも、URLを直接入力して自身が出品していない売却済み商品の商品購入ページへ遷移しようとすると、トップページに遷移する動画
https://gyazo.com/466400d2e6a123994c28c271c2b9d101

４
 ログイン状態の場合でも、URLを直接入力して自身が出品した商品の商品購入ページに遷移しようとすると、商品の販売状況に関わらずトップページに遷移する動画
https://gyazo.com/227ab235bb5031f534d8021bdc35390e　（売却済み）
https://gyazo.com/a3124179480e2bd67091a8516480fe54　（未売却）

５ 
ログアウト状態の場合は、URLを直接入力して商品購入ページに遷移しようとすると、商品の販売状況に関わらずログインページに遷移する動画
https://gyazo.com/1b9cb9119ee6fbe923d1c75e7c861f9a
 
６・７
売却済みの商品は、画像上に「sold out」の文字が表示される動画（商品一覧、詳細機能実装時に未実装であった場合）
https://gyazo.com/a6328a67cffd07c13bea2e52db210adc

８
 ログイン状態の場合でも、売却済みの商品には、「商品の編集」「削除」「購入画面に進む」ボタンが表示されない動画（商品詳細機能実装時に未実装であった場合）
https://gyazo.com/143046f15425206ca6e9f925b92e8a74　（出品者が本人）
https://gyazo.com/40fa2e3a22aa4ad8ead23457e41cde24　（出品者が他人）

９
 ログイン状態の場合でも、URLを直接入力して自身が出品した売却済み商品の商品情報編集ページへ遷移しようとすると、トップページに遷移する動画（商品情報編集機能実装時に未実装であった場合）
https://gyazo.com/6fc89d3675986d3f6ed331d516ffafc5

１０ 
テスト結果の画像
https://gyazo.com/b807a68d8e862f91bcf521d1d054f4ca

よろしくお願いします。

DDT1129 committed:
ordersモデルの生成からビュー表示までの設定

DDT1129 committed:
orderモデルの生成、モデル間のアソシエーションの設定、orderコントローラーindexアクションの訂正

DDT1129 committed:
Formオブジェクトの実装

DDT1129 committed:
orderのルート、バリデーション、アソシエーション、コントローラの修正とview表示（商品名等の登録商品の各項目）

newアクションは不要であったため、ルートからアクションまで削除

アソシエーションを組み忘れて、フォームから送られてくるパラメーターが空の状態になっていたため修正。

ルボコップの実施

DDT1129 committed:
Formオブジェクトモデルの単体テストの作成

DDT1129 committed:
rubocopの実施

DDT1129 committed:
orderコントローラーの同記述をbefore_actionで整理

DDT1129 committed:
クレジットカード決済機能の実装１

DDT1129 committed:
クレジットカード決済機能実装２（サーバーサイドへのトークン送付）

DDT1129 committed:
クレジットカード決済機能実装３（決済機能のサーバーサイド実装）

DDT1129 committed:
クレジットカード決済機能実施４（環境変数の設定）

DDT1129 committed:
画像上に「sold out」の文字表示（商品一覧,詳細のページ）

DDT1129 committed:
条件分岐（商品、出品者、ログイン状態）

商品：売却済みかどうか
出品者：自身か他人か
ログイン：有無

DDT1129 committed:
rubocopの実施

DDT1129 committed:
条件分岐（詳細ページのへの遷移）

DDT1129 committed:
rubocopの実施

CHURO-MAU reviewed (commented):
レビュー依頼ありがとうございます！

コメントいたしましたので、ご確認お願いいたします。
修正後は、再度レビュー依頼フォームのご提出をお願いいたします。

DDT1129 commented on spec/models/order_address_spec.rb:
```diff
@@ -0,0 +1,118 @@
+        expect(@order_address.errors.full_messages).to include("Prefecture can't be blank")
+      end
+      it '都道府県が---では登録できない' do
+        @order_address.prefecture_id = '---'
```
修正しました。

DDT1129 committed:
order_addressモデルのバリデーション、テストの修正

DDT1129 commented on spec/factories/order_addresses.rb:
```diff
@@ -0,0 +1,15 @@
+    building           { Faker::Address.street_address }
+    phone              { Faker::Number.between(from: 1_000_000_000, to: 99_999_999_999) }
+
+    user_id            { Faker::Number.non_zero_digit }
+    market_id          { Faker::Number.non_zero_digit }
```
修正しました。
order_address_spec.rbファイルでそれぞれインスタンス生成し、Formオブジェクトのインスタンス変数にid値を入れました。

DDT1129 commented on spec/models/order_address_spec.rb:
```diff
@@ -0,0 +1,118 @@
+require 'rails_helper'
+
+RSpec.describe OrderAddress, type: :model do
+  before do
+    @order_address = FactoryBot.build(:order_address)
+  end
```
修正しました。
order_address_spec.rbファイルでそれぞれインスタンス生成し、Formオブジェクトのインスタンス変数にid値を入れました。その後、モデル単体テストの結果を確認したところ異常ありませんでした。

DDT1129 committed:
FacroryBotの修正

・インスタンスの生成（user_id,market_idの生成のため）
・外部キーカラムのデータに直接数値を設定しない（テーブル間（主、外部キー間）データの一貫性の維持のため）
・config/environments/test.rbに追記（beforeブロック内の処理が各テスト実行前に済まないことによるエラーを解消するため）

DDT1129 commented:
修正しました。
よろしくおねがいします。

# 動作確認の写真
テスト結果の画像
https://gyazo.com/91dbabfa59b306eed66cb744aa7d2e73

minachanchan reviewed (commented):
レビューのご依頼ありがとうございます。

実装内容を確認させていただきました。
**LGTMとさせていただきます。実装お疲れ様でした！**

このブランチをマージし、masterブランチ（もしくはmainブランチ）でプルしましょう。
内容が反映されていることを確認し、次の実装・学習に入りましょう！
引き続き応援しております！
